src/display/api.js

```diff
@@ -1510,24 +1516,36 @@ var WorkerTransport = (function WorkerTransportClosure() {
 
       messageHandler.on('NeedPassword',
                         function transportNeedPassword(exception) {
-        var loadingTask = this.loadingTask;
+        var error = new PasswordException(exception.message, exception.code);
+
         if (loadingTask.onPassword) {
-          return loadingTask.onPassword(updatePassword,
-                                        PasswordResponses.NEED_PASSWORD);
+          this._pendingException = error;
+          try {
+            loadingTask.onPassword(updatePassword,
+                                   PasswordResponses.NEED_PASSWORD);
+            this._pendingException = null;
+            return;
+          } catch (ex) { }
```

Comment:
I'm thinking we need to re-wrap the original exception here, e.g. at getDocument promise it will appear as InvalidPDFException but additional message from original exception.